core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,226 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ *
+ *
+ */
+// Discussion point. Using java.io.Closeable here so we can inherit AutoCloseable where its available.
+public final class Context implements Closeable {
+
+  /**
+   * The logical root context which is {@link #current()} if no other context is bound. This context
+   * is already closed and will not retain any listeners.
+   */
+  public static final Context ROOT = new Context();
+  static {
+    ROOT.close();
```

Comment:
Root being closed be default seems quite troublesome to use. It means you can't trust your context's isClosed to tell you useful information.

Since having close() on the root is problematic, could we use a different type and have `public CloseableContext withClose()`. That also makes listeners _much_ more efficient as every child context doesn't need to listen to its parent. As it stands today if you have a chain of 10 contexts and someone calls withListener, _every_ context in the chain has to create an object and hook it to its parent, since each context has its own close lifetime.